### PR TITLE
"default" dictionary key support. fixed #116

### LINKED ISSUE #116: Should be able to have dictionary key "default"
I think it should be possible to have a dictionary key named "default". The default for the dictionary should not pollute the namespace for keys.
Example:

```
delimiters "$","$"
d1 ::= [
  "default": "A",
  default: "B"
]
main() ::= <<
$d1.default$
$d1.foobar$
>>
```

Expected output

```
A
B
```

Actual

```
B
B
```

I also tried $d1.("default")$ but got the same result.

### PR DESCRIPTION
**Pull request by KvanTTT**
Fixes #116

### CODE REVIEW

File: src/org/stringtemplate/v4/STGroup.java [hunk 1]
```diff
@@ -80,9 +80,11 @@
 		TEMPLATE_FILE_EXTENSION = ".st";
 	}
 
-    /** When we use key as a value in a dictionary, this is how we signify. */
-    public static final String DICT_KEY = "key";
-    public static final String DEFAULT_KEY = "default";
+    /** When we use key as a value in a dictionary, this is how we signify.
+     *  Magic suffix is used to distinguish "default" string literal and "default" key
+     *  End part is the date of this suffix adding. */
+    public static final String DICT_KEY = "key_st4_magic_suffix_2016/01/10";
```
Comment (sharwell) [inline src/org/stringtemplate/v4/STGroup.java]: The change to `DICT_KEY` shouldn't be necessary, as the object is [compared by reference](https://github.com/antlr/stringtemplate4/blob/cb00ede2bd9ffbdbead442a7cda8c277ffdec8b9/src/org/stringtemplate/v4/misc/MapModelAdaptor.java#L52), not by value.

Comment (KvanTTT) [inline src/org/stringtemplate/v4/STGroup.java]: Ok. What do you think about appended "magic suffix" (actually postfix). Maybe you can propose a better string?

File: src/org/stringtemplate/v4/STGroup.java [hunk 2]
```diff
@@ -80,9 +80,11 @@
...
+    public static final String DEFAULT_KEY = "default_st4_magic_suffix_2016/01/10";
```
Comment (sharwell) [inline src/org/stringtemplate/v4/STGroup.java]: :exclamation: The string you use here is *less likely* to cause a problem, but all it does is change the magic string which breaks dictionary functionality. Use the following instead:

```java
public static final String DEFAULT_KEY = "\0";
```

This single character string is the shortest (and thus fastest to look up in a map) string which is not representable in the StringTemplate language for dictionaries.

Comment (KvanTTT) [inline src/org/stringtemplate/v4/STGroup.java]: I tried this, but unfortunately is does not solve the problem. See [this test on gist](
https://gist.github.com/KvanTTT/6cac1d75061995e96a71).
Escaped chars are allowed in keys (actually key can be any string due to escaping).

Comment (KvanTTT) [inline src/org/stringtemplate/v4/STGroup.java]: What about this test case:
```java

public static final String DEFAULT_KEY = "\0"; // If replace with "asdf", for example, test will be passed.

...

@Test public void testDictDefaultStringAsKey() throws Exception {
    String templates =
            "typeInit ::= [\"\0\":\"foo\", default:key] "+newline+
            "var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
            ;
    writeFile(tmpdir, "test.stg", templates);
    STGroup group = new STGroupFile(tmpdir+"/"+"test.stg");
    ST st = group.getInstanceOf("var");
    st.add("type", "\0");
    st.add("name", "x");
    String expecting = "\0 x = foo;";
    String result = st.render();
    assertEquals(expecting, result); // expecting: " x = foo;" result: " x = ;"
}
```
Of course it is unlikely but possible. Actually key can be any string due to escaping.

**Comment by sharwell** [comment]
I meant there is no escape sequence in a .st or .stg file for this. You would need to include a literal NUL in the file, as \0 would not work as an escape.

File: src/org/stringtemplate/v4/STGroup.java (hunk 1 above)
Comment (sharwell) [inline src/org/stringtemplate/v4/STGroup.java]: I would actually use this:

```java
public static final String DICT_KEY = new String("key");
```

This guarantees that the string `DICT_KEY` is different from any other string running in the application, so when it gets compared by reference there is no problem.

**Pull request closed without merging**